Extracted iterator resolution

XedinUnknown opened the pull request:
Aims to incorporate solution for #3.

XedinUnknown committed:
Added dev dep on stringable interface

XedinUnknown committed:
Added autoload paths

XedinUnknown committed:
Added PHPS project files

XedinUnknown committed:
Aliased `develop` to `0.1.x-dev`

XedinUnknown committed:
Added test for `ResolveIteratorCapableTrait`

XedinUnknown committed:
Added `ResolveIteratorCapableTrait`

XedinUnknown committed:
Added test for `CountIterableCapableTrait`

XedinUnknown committed:
Added `CountIteratorCapableTrait`

XedinUnknown committed:
Fixed test

Had wrong expectation of default param

XedinUnknown committed:
Removed unnecessary condition

There's not an explicit return in all cases. This also brings
coverage to 100%.

mecha commented on src/NormalizeIteratorCapableTrait.php:
```diff
@@ -40,9 +40,8 @@ protected function _normalizeIterator($value)
-        if ($value instanceof Traversable) {
-            return $this->_createTraversableIterator($value);
-        }
+        // If not array them gotta be traversable
```
then*

mecha commented on src/ResolveIteratorCapableTrait.php:
```diff
@@ -0,0 +1,137 @@
+        $i = 0;
+        while (!$test($iterator) && $i < $limit) {
+            /* @var $iterator IteratorAggregate */
+            $_it = $iterator->getIterator();
```
Is `$iterator` guaranteed to be an `IteratorAggregate` here? Perhaps I missed something. To me it seems like it could be anything, and the only definitive thing it can be is `Traversable`.

XedinUnknown replied:
1. `Traversable` is not "anything".
2. It is one of two things, and if it's not an `Iterator` then it is an `IteratorAggregate`.
3. `getIterator()` returns a `Traversabl`. Go to point 2.

mecha replied:
The param is type hinted as `Traversable`.

It is copied into a temporary variable `$origIterator`.

In a `while` loop, it is tested against the `test` callback.

I don't see anything that says it's not an `Iterator`, and I don't see anything that says it _has_ to be an `IteratorAggregate`. I'm not saying this is the case. But I don't see anything that indicates otherwise. I'm just concerned about a potential loose assumption.

mecha commented on src/CountIterableCapableTrait.php:
```diff
@@ -34,17 +38,13 @@
+                    return $it instanceof Countable;
+                });
+            } catch (RootException $e) {
+                return;
```
Why returning nothing (i.e. `null`) when an exception is thrown? I would assume the exception should either bubble up, get wrapped, or at least document that `null` can be returned.

XedinUnknown replied:
Document where?

mecha replied:
Sorry. I didn't realize this was a closure. I see now that the return value is later checked against certain types.

mecha commented on src/ResolveIteratorCapableTrait.php:
```diff
@@ -0,0 +1,137 @@
+        $origIterator = $iterator;
+
+        // Default test function
+        if (!is_callable($test)) {
```
I would make this more strict. If `null`, use default. If not `null` and not a `callable`, throw.

mecha replied:
Otherwise, in the event of a bug in consuming code, this would "fail" silently if a non-callable argument is given.

XedinUnknown committed:
Added test for `IteratorAggregate`

This is because the test function is not guaranteed to test for
`Iterator` interface.
